### PR TITLE
Add `index.mode: time_series`

### PR DESCRIPTION
**Pull request by nik9000**
This adds a setting to enable time series mode that is hidden by the
`es.index_mode_feature_flag_registered` feature flag. All it does right
now is make sure you haven't configured index sorting and partitioning.
This gives us a place to "hang" all of our further work on time series
mode.

Time series mode will entirely take over index sorting. We don't believe
you'll ever be able to configure sorting yourself when you are in time
series mode. We don't expect time series mode to support index
partitioning when we first build it but we'd like to get there.

### CODE REVIEW

**Comment by nik9000** [comment]
This is a tiny tiny tiny part of the rebuild of #75638. But opening it is a good sign! I feel like I really can get to the meat of the rebuild soon.

**Review by mark-vieira — APPROVED** [review]
LGTM. I assume you ran a test build locally with `-Dbuild.snapshot=false` to see the tests were ok? We don't actually run in that configuration in PR or intake jobs.

**Comment by nik9000** [comment]
> LGTM. I assume you ran a test build locally with `-Dbuild.snapshot=false` to see the tests were ok? We don't actually run in that configuration in PR or intake jobs.

I hadn't. Let me give that a go.

**Comment by nik9000** [comment]
> I hadn't. Let me give that a go.

Well first time trying it everything crashed. I've stared with `--max-workers 2` to see how that goes.

**Comment by nik9000** [comment]
> Well first time trying it everything crashed. I've stared with `--max-workers 2` to see how that goes.

Yeah, the load average was 130 and the machine has 12 cores....

With `-max-workers 2` I'm getting load average of 10.

**Comment by nik9000** [comment]
After a few false starts I have some progress!
```
<====---------> 34% EXECUTING [50m 57s]
```

It's been a long, long, long time since I ran the entire suite locally.

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 1]
```diff
@@ -317,12 +321,58 @@
...
+            public void validate(IndexMode value, Map<Setting<?>, Object> settings) {
+                value.validateWithOtherSettings(settings);
+            }
+
+            @Override
+            public Iterator<Setting<?>> settings() {
+                return IndexMode.VALIDATE_WITH_SETTINGS.iterator();
+            }
+        },
+        Property.IndexScope
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/index/IndexSettings.java:362]: This should probably also have `Property.Final`.

Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:362]: Indeed `Property.Final` seems like that way to go here.

Comment (imotov) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:362]: Should `INDEX_ROUTING_PARTITION_SIZE_SETTING` be Final as well? It is not marked Final but it seems like it shouldn't be set after index creation.

**Review by imotov — APPROVED** [review]
LGTM.

File: server/src/main/java/org/elasticsearch/index/IndexSettings.java (hunk 1 above)
Comment (nik9000) [reply server/src/main/java/org/elasticsearch/index/IndexSettings.java:362]: I opened #77669 to talk about that.

**Comment by nik9000** [comment]
Lovely:
```
$  ./gradlew --max-workers 2 -Dbuild.snapshot=false -Dlicense.key="x-pack/license-tools/src/test/resources/public.key" -p qa check
...

BUILD FAILED in 1h 46m 30s
```

Failed overnight. `:qa:snapshot-based-recoveries:gcs:integTest` got an empty reply from the server. Checking it again locally.

I think, maybe, we need a way to make CI do this for me. Or we need some better way for me to do it locally.

**Comment by nik9000** [comment]
> Failed overnight. `:qa:snapshot-based-recoveries:gcs:integTest` got an empty reply from the server. Checking it again locally.

Didn't fail locally on second run, either with `build.snapshot=false` or without it. Gremlins killed my build.

**Comment by nik9000** [comment]
I've tested this a bunch locally and it looks ok. I've merged to master and manually kicked off a non-snapshot build using it. I'm going to wait to backport until it passes.

**Comment by nik9000** [comment]
I'm pulling this back out of 7.16 because we don't really have a chance of making a useful tsdb in time. So there isn't really any need to have the setting in that version.

**Comment by nik9000** [comment]
#78570 is the revert PR

**Pull request merged**